src/core/SampleBufferV2.cpp

```diff
@@ -0,0 +1,220 @@
+}
+
+int SampleBufferV2::numFrames() const
+{
+	return m_sampleData.size();
+}
+
+void SampleBufferV2::resample(const int oldSampleRate, const int newSampleRate)
+{
+	int dstFrames = static_cast<int>(static_cast<float>(numFrames()) / oldSampleRate * newSampleRate);
+	auto resampleBuf = std::vector<sampleFrame>(dstFrames);
+
+	int error;
+	SRC_STATE* state;
+	if ((state = src_new(SRC_LINEAR, DEFAULT_CHANNELS, &error)) != nullptr)
+	{
+		SRC_DATA srcData;
+		srcData.data_in = m_sampleData.data()->data();
+		srcData.input_frames = numFrames();
+		srcData.data_out = resampleBuf.data()->data();
+		srcData.output_frames = dstFrames;
+		srcData.src_ratio = static_cast<double>(newSampleRate) / oldSampleRate;
+		srcData.end_of_input = 1;
+
+		error = src_process(state, &srcData);
+		src_delete(state);
+	}
+
+	if (error != 0)
+	{
+		throw std::runtime_error(std::string("An error occurred when resampling: ") + src_strerror(error) + '\n');
+	}
+
+	m_sampleData = std::move(resampleBuf);
+}
+
+void SampleBufferV2::loadFromAudioFile(const std::filesystem::path& audioFilePath)
+{
+	SF_INFO sfInfo;
+	sfInfo.format = 0;
+
+	auto sndFileDeleter = [](SNDFILE* ptr) { sf_close(ptr); };
+	#ifdef LMMS_BUILD_WIN32
+		auto sndFile = std::unique_ptr<SNDFILE, decltype(sndFileDeleter)>(
+			sf_wchar_open(audioFilePath.c_str(), SFM_READ, &sfInfo), sndFileDeleter);
+	#else
+		auto sndFile = std::unique_ptr<SNDFILE, decltype(sndFileDeleter)>(
+			sf_open(audioFilePath.c_str(), SFM_READ, &sfInfo), sndFileDeleter);
+	#endif
+
+	if (!sndFile) { throw std::runtime_error("Failed to open audio file: " + std::string{sf_strerror(sndFile.get())}); }
+
+	auto numSamples = sfInfo.frames * sfInfo.channels;
+	auto samples = std::vector<float>(numSamples);
+	auto samplesRead = sf_read_float(sndFile.get(), samples.data(), numSamples);
+
+	if (samplesRead != numSamples)
+	{
+		throw std::runtime_error("Failed to read audio samples: samplesRead != numSamples");
+	}
+
+	m_sampleData = std::vector<sampleFrame>(sfInfo.frames);
+	m_originalSampleRate = sfInfo.samplerate;
+	m_filePath = audioFilePath;
+
+	for (sf_count_t frameIndex = 0; frameIndex < sfInfo.frames; ++frameIndex)
+	{
+		m_sampleData[frameIndex][0] = samples[frameIndex * sfInfo.channels];
+		m_sampleData[frameIndex][1] = samples[frameIndex * sfInfo.channels + (sfInfo.channels > 1 ? 1 : 0)];
+	}
+
+	auto audioEngineSampleRate = Engine::audioEngine()->processingSampleRate();
+	if (sfInfo.samplerate != static_cast<int>(audioEngineSampleRate))
+	{
+		resample(sfInfo.samplerate, audioEngineSampleRate);
+	}
+}
+
+void SampleBufferV2::loadFromDrumSynthFile(const std::filesystem::path& drumSynthFilePath)
+{
+	auto dsFilePathStr = drumSynthFilePath.native();
+	auto ds = DrumSynth();
+	auto samples = std::make_unique<int16_t>();
+	auto samplesRawPtr = samples.get();
+
+	// TODO: Remove QString::fromStdString when Qt is removed from DrumSynth
+	int numSamples = ds.GetDSFileSamples(QString::fromStdString(dsFilePathStr), samplesRawPtr,
```

Comment:
I've changed `DrumSynth` to use `QFile` because `std::fstream` was unable to handle Unicode paths correctly on Windows, prior to C++17. It's possible to revert to `std::fstream` now, but that will conflict with #6078, so I'm not sure what is the best way to do this.

On Windows, `QString::fromStdWString` should be used instead, if we're going to keep using `QString` here.

Reply:
I am not sure what to do about this either. We are in the midst of removing Qt from the core, so I don't feel comfortable using ``QString`` in general. However, with the #6078 conflict you shown above, I am at a loss for course of action except waiting untill things are more settled.

Reply:
I have added ``QString`` to and from ``std::experimental::filesystem::path`` conversion functions within ``SampleBufferV2``. That way, if a file path needs to be a ``QString``, the conversion can occur respecting portability with Windows and Unix. Ideally, these functions wouldn't be a part of the core, but I wasn't sure where to put them but in ``SampleBufferV2``.

Reply:
Hopefully my final update on this: I decided to keep the file paths as a ``QString``, just because it makes interfacing with the already existing Qt interface a lot easier. We can always switch when we decide to fully remove ``QString`` and are ready to use ``std::filesystem::path``. Feel free to resolve this if that's okay.